AAP-36737 - Improvements to the authentication mapping content (#3010)
ansible/aap-docs: downstream/modules/platform/con-gw-understanding-authenticator-mapping.adoc

:_mod-docs-content-type: CONCEPT

[id="gw-understanding-authenticator-mapping"]

= Understanding authenticator mapping

Authentication:: validates a user’s identity, typically through a username and password or a trust system.

> Authorization:: determines what an authenticated user can do once they are authenticated. 

Reviewer: Just a nit: 
```suggestion
Authorization:: determines what an authenticated user can do after they are authenticated. 
```
I'll leave this one up to you. Here's the IBM guideline applicable [here](https://www.ibm.com/docs/en/ibm-style?topic=word-usage#o): 
"(Use with caution) once adverb
Use only to mean one time. Do not use as a conjunction to mean after or when."

Author: Leaving the "Once" phrase because the meaning is that after you enter the information one time, it becomes available in the drop-down going forward.

In {PlatformNameShort}, authenticators manage authentication, validating users and returning details such as their username, first name, email, and group memberships (for example, LDAP groups). Authorization comes from the authenticator’s associated maps.

> During the authentication process, once a user has authenticated, the authorization system starts with a default set of permissions in memory. Then sequentially, the authenticator maps are processed and adjust permissions based on their trigger conditions. Once all the authenticator's maps are processed, the in-memory representation of the users permissions are reconciled with the users existing permissions. 

Reviewer: Same as above (use "once" to mean "one time", per IBM; not a replacement for "after" or "when"). Will leave this up to you.

```suggestion
During the authentication process, after a user has authenticated, the authorization system starts with a default set of permissions in memory. Then sequentially, the authenticator maps are processed and adjust permissions based on their trigger conditions. When all the authenticator's maps are processed, the in-memory representation of the user's permissions are reconciled with the user's existing permissions. 
```
Assuming users in this case is possessive and singular. If I've assumed incorrectly, please disregard.

Author: Fixed.

> For example, let's simplify the in-memory representation of the default permissions as follows:

Reviewer: ```suggestion
For example, simplify the in-memory representation of the default permissions as follows:
```
Avoid first person (IBM style guide, [Person > First person](https://www.ibm.com/docs/en/ibm-style?topic=grammar-verbs#person). See examples under first person, specifically the example "Let's start by planning the home page...":

Author: Fixed.

-----
Access allowed = True
Superuser permission = Undefined
Admin of teams = None
-----

> And, let’s say the maps which need to be processed are processed in the following order:

Reviewer: Similar to above, removing the first person style: 

```suggestion
And, you might have maps that need to be processed in the following order:
```

Author: Fixed.

. *Allow* rule set to never
. *Allow* rule based on group
. *Superuser* rule based on user attributes
. *Team* admin rule based on user group

> The first *Allow* map, set to never, denies access to the system and the in memory representation would now look like:

Reviewer: ```suggestion
The first *Allow* map, set to never, denies access to the system and the in-memory representation looks like:
```
IBM style guide, usage of "would" (verb), https://www.ibm.com/docs/en/ibm-style?topic=word-usage#w:
(Use with caution) **would** _verb_
Use only to refer to an unfulfilled condition or hypothesis and to express the future in the past, for “The command was rejected because it would have resulted in the ClusterName attribute being blank.” Do not use as an auxiliary verb, for example, “You would need administrator authority to create an app.” Instead, use the present tense, for example “You need administrator authority to create an app.”

Author: Fixed.

[subs=+quotes]
-----
Access allowed = *False*
Superuser permission = Undefined
Admin of teams = None
-----

However, if the user matches the second *Allow* map (the group-based allow), the permissions change to the following:

[subs=+quotes]
-----
Access allowed = *True*
Superuser permission = Undefined
Admin of teams = None
-----

> Where the user is subsequently granted access to Ansible Automation Platform because they have the required groups. 

Reviewer: Use attribute: 

```suggestion
Where the user is subsequently granted access to {PlatformNameShort} because they have the required groups. 
```

Author: Fixed.

> Next, the *Superuser* map checks user attributes. If no match is found, it does not revoke existing permissions by default. Therefore, the permissions remain the same as the results from our last map:

Reviewer: ```suggestion
Next, the *Superuser* map checks user attributes. If no match is found, it does not revoke existing permissions by default. Therefore, the permissions remain the same as the results from your last map:
```
Keep content user focused ("you")

Author: Fixed.

[subs=+quotes]
-----
Access allowed = True
Superuser permission = *Skipped*
Admin of teams = None
-----

To revoke superuser access, you can select the *Revoke* option on the *Superuser* map. That way, when the user does not meet the attribute criteria, the permissions update to False like in the following:

[subs=+quotes]
-----
Access allowed = True
Superuser permission = *False*
Admin of teams = None
-----

The final *Team* map checks the user’s groups coming from the authenticator for admin access on the team “My Team”. If the user has the required group, the permissions update to the following:

[subs=+quotes]
-----
Access allowed = True
Superuser permission = False
Admin of teams = “*My Team*”
-----

If the user lacks the required group, permissions remain unchanged unless the *Revoke* option has been selected on the map, in which case permissions update to the following:

[subs=+quotes]
-----
Access allowed = True
Superuser permission = False
Admin of teams = *Revoke admin of “My Team”*
-----
After processing all maps in the order defined, the final permissions reconcile, updating the user’s access based on the map rules.

> In summary, authenticators authenticate users and delegate system authorization to the authenticator maps defined for the authenticator. Authenticator maps are executed in order creating an in memory representation of the users permissions which get reconciled with the users actual permissions once all maps are executed.

Reviewer: Just thinking of ways to be less redundant with the use of "authenticate" and "authenticator"

```suggestion
In summary, authenticators validate users and delegate system authorization to the authenticator maps. Authenticator maps are executed in order creating an in-memory representation of the users permissions which get reconciled with the users actual permissions after all maps are executed.
```

Author: Fixed.

By default, authenticator maps return either *ALLOW* or *SKIPPED*.

> ALLOW:: Means that a match is detected and the platform should grant  the user access to  the corresponding role or permission (such as, superuser, or team member).

Reviewer: Nit

```suggestion
ALLOW:: Means that a match is detected and the platform should grant  the user access to the corresponding role or permission (such as, superuser, or team member).
```

Reviewer: @dcdacosta  - I wasn't sure if the spaces here were updated.

> SKIPPED:: Means that the user did not match the trigger in the map and , the platform skips processing this map and continues to check the remaining maps. This is useful if you want to grant users additional permissions in the system without having to change the authenticator maps.

Reviewer: ```suggestion
SKIPPED:: Means that the user did not match the trigger in the map, and the platform skips processing this map and continues to check the remaining maps. This is useful if you want to grant users additional permissions in the system without having to change the authenticator maps.
```

Author: Fixed.
